backstory/src/main/java/com/google/sps/servlets/AnalyzeImageServlet.java

```diff
@@ -0,0 +1,211 @@
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Prototype Servlet for serving the POST request for the analyze image resource. It uses the Vision
+ * API to analyze the images.
+ */
+@WebServlet("/analyze-image")
+public class AnalyzeImageServlet extends HttpServlet {
+  /**
+   * {@inheritDoc}
+   *
+   * Expecting a post request from Blobstore containing the data fields from the image-upload
+   * form. After having gone through Blobstore, the request will include the image uploaded. The
+   * form in the HTML will connect to the Blobstore URL, which encodes the image and then
+   * redirects the request to this Url. In this doPost, the image is analyzed with Vision API,
+   * and the analysis along with the image URL will be sent to datastore for permanent storage.
+   */
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    // Get the URL of the image that the user uploaded to Blobstore.
+    final String imageUrl = getUploadedFileUrl(request, "image-upload");
+
+    // Get the raw byte array representing the image from Blobstore
+    final byte[] bytes = getBlobBytes(request, "image-upload");
+
+    if (bytes == null || imageUrl == null) {
+      // Redirect back to the HTML page.
+      response.sendError(400, "Please upload a valid image.");
+
+    } else {
+      // Gets the full label information from the image byte array by calling Vision API.
+      List<EntityAnnotation> labels = detectLabelsFromImageBytes(bytes);
+      Gson gson = new Gson();
+      Text labelsJsonArray = new Text(gson.toJson(labels));
+
+      // Add the input to datastore
+      Entity analyzedImageEntity = new Entity("analyzed-image");
+      analyzedImageEntity.setProperty("imageUrl", imageUrl);
+      analyzedImageEntity.setProperty("labelsJsonArray", labelsJsonArray);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(analyzedImageEntity);
+
+      // Redirect back to the HTML page.
+      response.sendRedirect("/vision-upload-prototype/vision-demo.html#image-upload");
+    }
+  }
+
+  /** Detects labels in the image specified by the image byte data by calling the Vision API. */
+  private List<EntityAnnotation> detectLabelsFromImageBytes(byte[] bytes) throws IOException {
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    List<EntityAnnotation> labels;
+
+    Image img = Image.newBuilder().setContent(ByteString.copyFrom(bytes)).build();
+    Feature feat = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
```

Comment:
Avoid non-standard abbreviations if possible, to improve readability. Applies to `img` and `feat`.

This is from the JS style guide but the principle also applies here:
"Do not worry about saving horizontal space as it is far more important to make your code immediately understandable by a new reader. Do not use abbreviations that are ambiguous or unfamiliar to readers outside your project, and do not abbreviate by deleting letters within a word."
https://google.github.io/styleguide/jsguide.html#naming-rules-common-to-all-identifiers